Migrate rtt_tf to tf2

francisco-miguel-almeida opened the pull request:
This branch ports rtt_tf to be now based on tf2. It also allows subscribing to static transforms as input.
As requested in issue #68.

francisco-miguel-almeida committed:
refactor component to be implemented based on tf2::BufferCore

francisco-miguel-almeida committed:
InputPort port_tf_static_in is connected to /tf_static topic

francisco-miguel-almeida committed:
add rtt_tf2_msgs typekit for support of tf2 types in RTT

francisco-miguel-almeida committed:
update rtt_tf tests to load any useful plugins automatically

francisco-miguel-almeida committed:
Update README by referring to tf2.

meyerj commented on rtt_tf/README.md:
```diff
@@ -31,6 +31,9 @@ and then connect it with whichever components you intend to have use it:
 import("rtt_ros");
+import("rtt_std_msgs");
+import("rtt_geometry_msgs");
+import("rtt_tf2_msgs");
```
Not needed.

```
import("rtt_ros");
```
provides the `ros.import` operation and
```
ros.import("rtt_tf")
```
then imports all packages recursively which are listed in `<plugin_depend>` entries in the respective `package.xml` files.

meyerj commented on rtt_tf/rtt_tf-component.cpp:
```diff
@@ -49,22 +49,78 @@
+    }
+    return composite + frame_name;
+  }
+}
```
I think it would be fine to depend on `tf` and use `tf::resolve()` from there.
If not, rename the method to `tf::resolve()` if it is an exact copy?

meyerj commented on rtt_tf/rtt_tf-component.cpp:
```diff
@@ -49,22 +49,78 @@
-    TaskContext(name, PreOperational), 
-    tf::Transformer(true, ros::Duration(Transformer::DEFAULT_CACHE_TIME)),
-    prop_cache_time(Transformer::DEFAULT_CACHE_TIME),
+    TaskContext( name, PreOperational ),
```
nit: avoid adding spaces after open and close parentheses

This style is sometimes used within RTT, but in general rtt_ros_integration should follow the [ROS C++ style guide](http://wiki.ros.org/CppStyleGuide).

meyerj commented on rtt_tf/rtt_tf-component.cpp:
```diff
@@ -49,22 +49,78 @@
     this->addProperty("tf_prefix", prop_tf_prefix);
     this->addEventPort("tf_in", port_tf_in);
+    this->addEventPort("tf_static_in", port_tf_static_in);
     this->addPort("tf_out", port_tf_out);
```
Please also add an output port for static transforms and two new operations `broadcastStaticTransform` and `broadcastStaticTransforms`.

meyerj commented on rtt_tf/rtt_tf-component.cpp:
```diff
@@ -109,125 +165,148 @@ namespace rtt_tf
-    // Update the tf::Transformer prefix
-    tf_prefix_ = prop_tf_prefix;
+    // no prefix to update in tf2::BufferCore (see #68)
+    //tf_prefix_ = prop_tf_prefix;
```
nit: Remove obsolete code.

meyerj commented on rtt_tf/rtt_tf-component.cpp:
```diff
@@ -109,125 +165,148 @@ namespace rtt_tf
+    cp.transport = 3;
+    cp.name_id = "/tf_static";
+
+    return (port_tf_static_in.createStream(cp_static) && port_tf_in.createStream(cp) && port_tf_out.createStream(cp));
```
nit: Call `cleanupHook()` in case of an error and disconnect all ports there. Just for the sake of a well-defined configure/cleanup behavior.

meyerj commented on rtt_tf/rtt_tf-component.cpp:
```diff
@@ -109,125 +165,148 @@ namespace rtt_tf
   }
 
-  void RTT_TF::updateHook()
+  void RTT_TF::internalUpdate(tf2_msgs::TFMessage& msg, RTT::InputPort<tf2_msgs::TFMessage>& port)
```
No need to pass `msg` from the outside. It can be a local variable.

You can pass a "slot function" (user callback) in [addEventPort](https://orocos-toolchain.github.io/rtt/toolchain-2.9/api/html/classRTT_1_1TaskContext.html#af5d3991d677caee96a58bee04304afe6) and bind this function directly, without the need to implement/override the `updateHook()`:

```cpp
void RTT_TF::tfCallback(RTT::base::PortInterface *_port) {
  tf::tfMessage msg_in;
  RTT::InputPort<tf2_msgs::TFMessage> port
      = dynamic_cast<RTT::InputPort<tf2_msgs::TFMessage> >(_port);
  const bool is_static = (port == &port_tf_static_in);

  while (port.read(msg) == NewData) {
    // ...
    this->setTransform(trans, authority, is_static);
  }
```

meyerj commented on rtt_tf/rtt_tf-component.cpp:
```diff
@@ -109,125 +165,148 @@ namespace rtt_tf
 #if ROS_VERSION_MINIMUM(1,11,0)
-            this->setTransform(trans);
+          const std::string authority = "default_authority";
+          this->setTransform(msg.transforms[i], authority);
```
Set optional `is_static` flag if the transform was received on `port_tf_static_in`. See example code snippet above.

meyerj commented on rtt_tf/rtt_tf-component.cpp:
```diff
@@ -109,125 +165,148 @@ namespace rtt_tf
+// tf2::BufferCore (see #68) has a non-defaulted authority argument,
+//  but there is no __connection_header to extract it from.
+// Workaround: use the same default value as tf::Transform did.
 #if ROS_VERSION_MINIMUM(1,11,0)
```
You can remove the `#else` branch. `msg.__connection_header` [has been removed in indigo](http://wiki.ros.org/indigo/Migration#gencpp). There is no way at the moment to access the connection header for a message read from an Orocos input port (hence `"default_authority"`).

Maybe rename to `"default_authority"` to `"unknown_authority"`?

meyerj commented on rtt_tf/rtt_tf-component.cpp:
```diff
@@ -109,125 +165,148 @@ namespace rtt_tf
+    }
+  }
+
+  void RTT_TF::updateHook()
```
Remove in favor of `RTT_TF::tfCallback()` registered as a user callback for both event ports (see example above).

meyerj commented on rtt_tf/rtt_tf-component.cpp:
```diff
@@ -109,125 +165,148 @@ namespace rtt_tf
+
+  void RTT_TF::stopHook()
+  {
+  }
```
No need to overwrite trivial `startHook()` and `stopHook()`.

meyerj commented on rtt_tf/rtt_tf-component.cpp:
```diff
@@ -109,125 +165,148 @@ namespace rtt_tf
+
+  void RTT_TF::cleanupHook()
+  {
+  }
```
nit: disconnect ports here, the opposite of what is done in `configureHook()`

meyerj commented on typekits/rtt_tf2_msgs/package.xml:
```diff
@@ -0,0 +1,52 @@
+  <version>2.9.1</version>
+  <description>
+
+    Provides an rtt typekit for ROS nav_msgs messages.
```
nit: -> `tf2_msgs`

You can automatically generate this package by running
```
rosrun rtt_roscomm create_rtt_msgs tf2_msgs
```

meyerj commented on typekits/rtt_tf2_msgs/package.xml:
```diff
@@ -0,0 +1,52 @@
+  <run_depend>tf2_msgs</run_depend>
+
+  <run_depend>rtt_tf</run_depend>
+  <build_depend>rtt_tf</build_depend>
```
The typekit package cannot depend on `rtt_tf`. `rtt_tf` should depend on `rtt_tf2_msgs`.

meyerj committed:
rtt_tf2_msgs: update package by using create_rtt_msgs script from rtt_roscomm

meyerj commented on rtt_tf/rtt_tf-component.cpp:
```diff
@@ -109,125 +165,148 @@ namespace rtt_tf
+      const std::string& target,
+      const std::string& source) const
+  {
+    return tf2::BufferCore::lookupTransform(target, source, getCommonTime(target, source));
```
Simply pass `ros::Time()` as the last argument (lookup latest available transform) to [lookupTransform()](http://docs.ros.org/kinetic/api/tf2/html/classtf2_1_1BufferCore.html#abcda3c653aba238f8fdd4b903c4bac7b). I think the internal implementation will be more efficient than to lookup the latest common time first.

meyerj commented on rtt_tf/rtt_tf-component.cpp:
```diff
@@ -109,125 +165,148 @@ namespace rtt_tf
-    ros::Time common_time;
-    this->getLatestCommonTime(source, target, common_time,NULL);
-    return static_cast<tf::Transformer*>(this)->canTransform(target, source, common_time);
+    return tf2::BufferCore::canTransform(target, source, getCommonTime(target, source));
```
`return tf2::BufferCore::canTransform(target, source, ros::Time())`

meyerj commented on rtt_tf/rtt_tf-component.cpp:
```diff
@@ -109,125 +165,148 @@ namespace rtt_tf
-    this->getLatestCommonTime(source, target, common_time,NULL);
-    return static_cast<tf::Transformer*>(this)->canTransform(target, source, common_time);
+    return tf2::BufferCore::canTransform(target, source, getCommonTime(target, source));
   }
```
add `canTransformAtTime()` for completeness?

meyerj commented on rtt_tf/rtt_tf-component.hpp:
```diff
@@ -2,36 +2,44 @@
+        tf2_msgs::TFMessage& msg,
+        RTT::InputPort<tf2_msgs::TFMessage>& port);
+
+    ros::Time getCommonTime(
```
Rename to `getLatestCommonTime()` for consistency with [tf2::BufferCore::_getLatestCommonTime()](http://docs.ros.org/kinetic/api/tf2/html/classtf2_1_1BufferCore.html#af20f2a5ad43089c8cd3db1b04badb194)?

meyerj reviewed (changes requested):
Looks good. I only have a few comments and requests:

- Static transforms have to be added as such to `tf2::BufferCore::setTransform()`.
- I suggest to add support for broadcasting static transforms, too.
- Remove commented/obsolete code

Not covered by comments below:
- `rtt_tf` needs to depend on the new typekit package `rtt_tf2_msgs`. Include the generated header `rtt_tf2_msgs/typekit/Types.hpp` instead of `tf2_msgs/TFMessage.h`, which declares the template instantiations for `RTT::InputPort<tf2_msgs::TFMessage>` and `RTT::OutputPort<tf2_msgs::TFMessage>` as external. This pattern should always be used for types that have a typekit to reduce binary sizes and save compilation time.
- Update [rtt_tf/package.xml](https://github.com/orocos/rtt_ros_integration/blob/b66a3040681bc6c632a064eee9fdb47ce247ee77/rtt_tf/package.xml) and [rtt_tf/CMakeLists.txt](https://github.com/orocos/rtt_ros_integration/blob/b66a3040681bc6c632a064eee9fdb47ce247ee77/rtt_tf/CMakeLists.xml). At the moment it only works because `tf` depends on `tf2_ros` internally. We can keep the dependency to `tf` and still use `tf::resolve()` instead of copying the method.

When adding new or updating operations of the component, please do not forget to also update the public [tf_interface.h](https://github.com/orocos/rtt_ros_integration/blob/b66a3040681bc6c632a064eee9fdb47ce247ee77/rtt_tf/include/rtt_tf/tf_interface.h), too.

meyerj committed:
Merge branch 'toolchain-2.9' into rtt-migrate-tf2

francisco-miguel-almeida committed:
remove unnecessary import statements from README. The rtt_ros plugin loads a component that allows importing ROS packages following the dependency tree.

francisco-miguel-almeida committed:
rtt_tf: add dependencies on tf2, rtt_tf2_msgs to rtt_tf

francisco-miguel-almeida committed:
rtt_tf: add canTransformAtTime operation, remove code duplication and other minor cleanups

francisco-miguel-almeida committed:
rtt_tf: handle configureHook() failure

francisco-miguel-almeida committed:
rtt_tf: reduce code in RTT_TF::internalUpdate and differentiate static transform case; remove dead code

francisco-miguel-almeida committed:
rtt_tf: add output port for static transform and broadcastStaticTransform(s) operations

francisco-miguel-almeida commented:
Addressed most issues, but as we discussed earlier, some bigger changes should be addressed as a separate PR. See issue created to discuss this: https://github.com/orocos/rtt_ros_integration/issues/126
